command/format.go

```diff
@@ -15,86 +16,102 @@ import (
+	if !ok {
 		ui.Error(fmt.Sprintf("Invalid output format: %s", format))
 		return 1
 	}
-}
-
-func outputFormatJSON(ui cli.Ui, s *api.Secret) int {
-	b, err := json.Marshal(s)
-	if err != nil {
-		ui.Error(fmt.Sprintf(
-			"Error formatting secret: %s", err))
+	if err := formatter.Output(ui, data); err != nil {
+		ui.Error(fmt.Sprintf("Could not output secret: %s", err.Error))
 		return 1
 	}
-
-	var out bytes.Buffer
-	json.Indent(&out, b, "", "\t")
-	ui.Output(out.String())
 	return 0
 }
 
-func outputFormatJSONList(ui cli.Ui, s *api.Secret) int {
-	b, err := json.Marshal(s.Data["keys"])
-	if err != nil {
-		ui.Error(fmt.Sprintf(
-			"Error formatting keys: %s", err))
-		return 1
+type Formatter interface {
+	Output(ui cli.Ui, data interface{}) error
+}
+
+var Formatters = map[string]Formatter{
+	"json":  JsonFormatter{},
+	"table": TableFormatter{},
+	"yaml":  YamlFormatter{},
+}
+
+// An output formatter for json output of an object
+type JsonFormatter struct {
+}
+
+func (j JsonFormatter) Output(ui cli.Ui, data interface{}) error {
+	b, err := json.Marshal(data)
+	if err == nil {
+		var out bytes.Buffer
+		json.Indent(&out, b, "", "\t")
+		ui.Output(out.String())
 	}
+	return err
+}
 
-	var out bytes.Buffer
-	json.Indent(&out, b, "", "\t")
-	ui.Output(out.String())
-	return 0
+// An output formatter for yaml output format of an object
+type YamlFormatter struct {
 }
 
-func outputFormatYAML(ui cli.Ui, s *api.Secret) int {
-	b, err := yaml.Marshal(s)
-	if err != nil {
-		ui.Error(fmt.Sprintf(
-			"Error formatting secret: %s", err))
-		return 1
+func (y YamlFormatter) Output(ui cli.Ui, data interface{}) error {
+	b, err := yaml.Marshal(data)
+	if err == nil {
+		ui.Output(strings.TrimSpace(string(b)))
 	}
+	return err
+}
 
-	ui.Output(strings.TrimSpace(string(b)))
-	return 0
+// An output formatter for table output of an object
+type TableFormatter struct {
 }
 
-func outputFormatYAMLList(ui cli.Ui, s *api.Secret) int {
-	b, err := yaml.Marshal(s.Data["keys"])
-	if err != nil {
-		ui.Error(fmt.Sprintf(
-			"Error formatting secret: %s", err))
-		return 1
+func (t TableFormatter) Output(ui cli.Ui, data interface{}) error {
+	// TODO: this should really use reflection like the other formatters do
+	if s, ok := data.(*api.Secret); ok {
+		return t.OutputSecret(ui, s)
 	}
+	if s, ok := data.([]interface{}); ok {
+		return t.OutputSecretList(ui, s)
+	}
+	return errors.New("Cannot use the table formatter for this type")
+}
 
-	ui.Output(strings.TrimSpace(string(b)))
-	return 0
+func (t TableFormatter) OutputSecretList(ui cli.Ui, list []interface{}) error {
+	config := columnize.DefaultConfig()
+	config.Delim = "♨"
+	config.Glue = "\t"
+	config.Prefix = ""
+
+	input := make([]string, 0, 5)
+
+	input = append(input, "Keys")
+
+	keys := make([]string, 0, len(list))
+	for _, k := range list {
+		keys = append(keys, k.(string))
+	}
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		input = append(input, fmt.Sprintf("%s", k))
+	}
+
```

Comment:
Hi @rkuris ,

The warnings field is contained in the return JSON and is in `*api.Secret`, so in the JSON and YAML outputs it was automatically included (as structured data) since those basically just passed the data through. Table format wasn't the only format passing the warnings on, just the only one formatting them -- but this is true of any of the data that is returned.

Since the table output is what is generally assumed to be used for client-interactive use, we do still want the warnings being displayed. Thanks!